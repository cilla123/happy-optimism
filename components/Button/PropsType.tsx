export default interface PropsType {
  theme?: 'default' | 'primary' | 'success' | 'warning' | 'error';
  size?: 'xl' | 'lg' | 'sm' | 'xs';
  shape?: 'radius' | 'round' | 'circle';
  block?: boolean;
  bordered?: boolean;
  active?: boolean;
  focus?: boolean;
  disabled?: boolean;
  loading?: boolean;
  icon?: JSX.Element;
  onClick?: (e?: any) => void;
}
