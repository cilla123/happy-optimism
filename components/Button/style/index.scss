@import "./_button/_optimism-button-variable.scss";
@import "./_button/_optimism-button.scss";

/* 大小相关 */
@import './_button/_optimism-button-size.scss';

/* 风格相关 */
@import './_button/_optimism-button-style.scss';

/* 状态相关 */
@import './_button/_optimism-button-status.scss';
@import './_button/_optimism-button-bordered.scss';

/* Icon */
@import './_button/_optimism-button-icon.scss';

/* Button Group */
@import './_button/_optimism-button-group.scss';
