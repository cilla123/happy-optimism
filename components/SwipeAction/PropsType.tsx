export default interface PropsType {
  left?: Array<Object>;
  right?: Array<Object>;
  moveDistanceRatio: number;
  moveTimeSpan: number;
  animationDuration?: number;
  offset: number;
  autoClose?: boolean;
  disabled?: boolean;
  onOpen?: () => void;
  onClose?: () => void;
}
