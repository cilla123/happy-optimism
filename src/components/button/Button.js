import React, { Component, PureComponent } from 'react'
import setClass from 'classnames'
import Proptypes from 'prop-types'
import omit from 'lodash/omit'
// import Icon from 'icon'

const BLACK_LIST = [
    'type'
]

const BASE_LIST = ['href', 'target']

const BTN_BLACK_LIST = BASE_LIST.concat(BLACK_LIST)

const A_BLACK_LIST = BASE_LIST.concat(BLACK_LIST)

const wrapTextWithSpanTag = children => {
    return React.Children.map(children, child => {
        if (typeof child === 'string') {
            return <span>{child}</span>
        }
        return child
    })
}

export default class Button extends (PureComponent || Component) {
    
    static propTypes = {
        type: Proptypes.oneOf(['default', 'primary', 'success', 'danger', 'link']),
    }

    static defaultProps = {
        type: 'default'
    }

    constructor(props){
        super(props)
    }

    renderLink(){
        return (
            <a>11</a>
        )
    }

    renderButton(classNames, iconNode, wrapedChildren){
        const { component } = this.props
        const Node = component || 'button'
        const nodeProps = omit(this.props, BTN_BLACK_LIST)

        return (
            <Node 
                {...nodeProps}
            >
                { iconNode }
                { wrapedChildren }
            </Node>
        )
    }

    render(){
        const { type, prefix } = this.props
        const { className } = this.props
        const renderer = href || target ? 'renderLink' : 'renderButton'
        const classNames = setClass(
            {
                [`${prefix}-btn-${type}${outline ? '-outline' : ''}`] : type !== 'default',
            },
            `${prefix}-btn`,
            className
        )
        const iconNode = null;
        const wrapedChildren = wrapTextWithSpanTag(children)

        return this[renderer](classNames, iconNode, wrapedChildren)
    }

}
