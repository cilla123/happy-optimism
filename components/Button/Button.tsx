import React, { Component, PureComponent } from 'react'
import PropsType from './PropsType'
import PropTypes from 'prop-types'
import omit from 'lodash/omit'
import setClass from 'classnames'
import Icon from '../Icon'

const BLACK_LIST = [
    'type',
    'size',
    'htmlType',
    'block',
    'component',
    'disabled',
    'loading',
    'outline',
    'bordered',
    'icon',
    'className',
    'prefix',
]

const BTN_BLACK_LIST = ['href', 'target'].concat(BLACK_LIST)

const A_BLACK_LIST = ['href', 'target'].concat(BLACK_LIST)

const wrapTextWithSpanTag = children => {
    return React.Children.map(children, child => {
        return typeof child === 'string' ? <span>{ child }</span> : child
    })
}

class Button extends (PureComponent || Component)<PropsType, any> {

    static propTypes = {
        type: PropTypes.oneOf(['default', 'primary', 'success', 'danger', 'link']),
        size: PropTypes.oneOf(['large', 'medium', 'small']),
        htmlType: PropTypes.oneOf(['button', 'submit', 'reset']),
        className: PropTypes.string,
        block: PropTypes.bool,
        component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
        disabled: PropTypes.bool,
        loading: PropTypes.bool,
        outline: PropTypes.bool,
        bordered: PropTypes.bool,
        prefix: PropTypes.string,
        icon: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
    }
    
    static defaultProps = {
        type: 'default',
        size: 'medium',
        htmlType: 'button',
        className: '',
        block: false,
        disabled: false,
        loading: false,
        outline: false,
        bordered: true,
        prefix: 'happy-optimism',
    }

    constructor(props){
        super(props)
        this.handleClick = this.handleClick.bind(this)
    }

    /**
     * 处理点击事件
     */
    handleClick(event){
        if (this.props.disabled || this.props.loading) return

        if (this.props.onClick) {
            this.props.onClick(event)
        }
    }

    /**
     * 渲染 a 标签
     */
    renderLink(classNames, iconNode, wrapedChildren){
        const { component, disabled, loading, href = '', target } = this.props
        const Node = component ||  'a'
        const nodePorps = omit(this.props, A_BLACK_LIST)

        return (
            <Node
                {...(disabled || loading ? {} : {href, target})}
                {...nodePorps}
                className={classNames}
                onClick={ this.handleClick }
            >
                { iconNode }
                { wrapedChildren }
            </Node>
        )
    }

    /**
     * 渲染 button 标签
     */
    renderButton(classNames, iconNode, wrapedChildren) {
        const { component, disabled, loading, htmlType } = this.props;
        const Node = component || 'button';
        const nodeProps = omit(this.props, BTN_BLACK_LIST);

        return (
            <Node
                {...nodeProps}
                {...(htmlType ? { type: htmlType } : {})}
                className={classNames}
                disabled={disabled || loading}
                onClick={this.handleClick}
            >
                {iconNode}
                {wrapedChildren}
            </Node>
        )
    }

    render() {
        const {
            href,
            target,
            type,
            size,
            block,
            disabled,
            loading,
            outline,
            bordered,
            prefix,
            icon,
            children,
        } = this.props
        let renderer = href || target ? 'renderLink' : 'renderButton'
        let { className } = this.props
        let classNames = setClass(
            {
            [`${prefix}-btn-${type}${outline ? '-outline' : ''}`]:
                type !== 'default',
            [`${prefix}-btn-${size}`]: size !== 'medium',
            [`${prefix}-btn-block`]: block,
            [`${prefix}-btn-loading`]: loading,
            [`${prefix}-btn-disabled`]: disabled,
            [`${prefix}-btn-border-transparent`]: !bordered,
            },
            `${prefix}-btn`,
            className
        );
        
        const iconNode = icon ? <Icon type={icon} /> : null;        
        const wrapedChildren = wrapTextWithSpanTag(children);

        return this[renderer](classNames, iconNode, wrapedChildren);
    }

}

export default Button