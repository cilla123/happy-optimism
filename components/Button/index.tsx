import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Spinner from '../Spinner';

export interface ButtonProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Button extends PureComponent<ButtonProps, {}> {

  static defaultProps = {
    prefixCls: 'za-button',
    theme: 'default',
    block: false,
    bordered: false,
    active: false,
    focus: false,
    disabled: false,
    loading: false,
    onClick() {},
  };

  render() {
    const {
      prefixCls,
      className,
      theme,
      size,
      shape,
      icon,
      block,
      active,
      focus,
      bordered,
      disabled,
      loading,
      onClick,
      children,
      ...others,
    } = this.props;

    const classes = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
      [`shape-${shape}`]: !!shape,
      block,
      bordered,
      active,
      focus,
      disabled,
    });

    const iconRender = loading
      ? <Spinner className="rotate360" />
      : icon;

    const childrenRender = children && <span>{children}</span>;

    const contentRender = (!!icon || loading)
      ? <div className={`${prefixCls}-content`}>{iconRender}{childrenRender}</div>
      : childrenRender;

    return (
      <a
        role="button"
        aria-disabled={disabled}
        className={classes}
        onClick={e => !disabled && typeof onClick === 'function' && onClick(e)}
        onTouchStart={() => {}}
        {...others}
      >
        {contentRender}
      </a>
    );
  }
}
