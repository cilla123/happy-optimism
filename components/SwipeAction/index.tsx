import React, { PureComponent, cloneElement } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Events from '../utils/events';
import Drag from '../Drag';

export interface SwipeActionProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class SwipeAction extends PureComponent<SwipeActionProps, any> {

  static defaultProps = {
    prefixCls: 'za-swipeaction',
    left: [],
    right: [],
    moveDistanceRatio: 0.5,
    moveTimeSpan: 300,
    animationDuration: 300,
    offset: 10,
  };

  private isOpen: boolean = false;
  private touchEnd: boolean = true;
  private wrap;
  private left;
  private right;

  constructor(props) {
    super(props);
    this.state = {
      offsetLeft: 0,
    };
  }

  componentDidMount() {
    Events.on(document.body, 'touchstart', e => this.onCloseSwipe(e));
  }

  componentWillUnmount() {
    Events.off(document.body, 'touchstart', e => this.onCloseSwipe(e));
  }

  onDragStart = () => {
    if (this.isOpen) {
      this.touchEnd = false;
      this.close();
      return;
    }
    this.touchEnd = true;
  }

  onDragMove = (event, { offsetX, offsetY }) => {
    const { disabled } = this.props;

    if (!this.touchEnd || disabled) {
      return;
    }

    // 拖动距离达到上限
    const { offset } = this.props;
    const { offsetLeft } = this.state;
    const btnsLeftWidth = this.left && this.left.offsetWidth;
    const btnsRightWidth = this.right && this.right.offsetWidth;
    if (
      offsetX > 0 && (!btnsLeftWidth || offsetLeft >= btnsLeftWidth + offset) ||
      offsetX < 0 && (!btnsRightWidth || offsetLeft <= -btnsRightWidth - offset)
    ) {
      return false;
    }

    // 判断滚屏情况
    const distanceX = Math.abs(offsetX);
    const distanceY = Math.abs(offsetY);
    if (distanceX < 5 || (distanceX >= 5 && distanceY >= 0.3 * distanceX)) {
      return false;
    }

    event.preventDefault();

    this.doTransition({ offsetLeft: offsetX, animationDuration: 0 });
    return true;
  }

  onDragEnd = (_event, { offsetX, startTime }) => {
    const { animationDuration, moveDistanceRatio, moveTimeSpan } = this.props;
    const timeSpan = new Date().getTime() - startTime.getTime();
    const btnsLeftWidth = this.left && this.left.offsetWidth;
    const btnsRightWidth = this.right && this.right.offsetWidth;

    let distanceX = 0;
    let isOpen = false;

    if ((offsetX / btnsLeftWidth > moveDistanceRatio) || (offsetX > 0 && timeSpan <= moveTimeSpan)) {
      distanceX = btnsLeftWidth;
      isOpen = true;
    } else if ((offsetX / btnsRightWidth < -moveDistanceRatio) || (offsetX < 0 && timeSpan <= moveTimeSpan)) {
      distanceX = -btnsRightWidth;
      isOpen = true;
    }

    if (isOpen && !this.isOpen) {
      // 打开
      this.open(distanceX);
    } else if (!isOpen && this.isOpen) {
      // 关闭
      this.close();
    } else {
      // 还原
      this.doTransition({ offsetLeft: distanceX, animationDuration });
    }
  }

  onCloseSwipe = (e) => {
    if (!this.wrap) {
      return;
    }

    if (this.isOpen) {
      const pNode = ((node) => {
        while (node.parentNode && node.parentNode !== document.body) {
          if (node === this.wrap) {
            return node;
          }
          node = node.parentNode;
        }
      })(e.target);

      if (!pNode) {
        e.preventDefault();
        this.touchEnd = true;
        this.close();
      }
    }
  }

  open = (offsetLeft) => {
    const { animationDuration, onOpen } = this.props;
    this.isOpen = true;
    this.doTransition({ offsetLeft, animationDuration });
    if (typeof onOpen === 'function') {
      onOpen();
    }
  }

  close = () => {
    const { animationDuration, onClose } = this.props;
    this.isOpen = false;
    this.doTransition({ offsetLeft: 0, animationDuration });
    if (typeof onClose === 'function') {
      onClose();
    }
  }

  doTransition = ({ offsetLeft, animationDuration }) => {
    this.setState({ offsetLeft, animationDuration });
  }

  renderButton = (button, index) => {
    return cloneElement(button, {
      key: +index,
      onClick: (e) => {
        const onClick = button.props.onClick;

        if (onClick) {
          onClick(e);
        }

        if (this.props.autoClose) {
          this.close();
        }
      },
    });
  }

  renderButtons = (buttons, direction) => {
    const { prefixCls } = this.props;

    if (!buttons || buttons.length === 0) {
      return;
    }

    return (
      <div className={`${prefixCls}-actions-${direction}`} ref={(el) => { this[direction] = el; }}>
        {buttons.map(this.renderButton)}
      </div>
    );
  }

  render() {
    const { prefixCls, className, left, right, children } = this.props;
    const { offsetLeft, animationDuration } = this.state;
    const cls = classnames(`${prefixCls}`, className);
    const style = {
      WebkitTransitionDuration: `${animationDuration}ms`,
      transitionDuration: `${animationDuration}ms`,
      WebkitTransform: `translate3d(${offsetLeft}px, 0, 0)`,
      transform: `translate3d(${offsetLeft}px, 0, 0)`,
    };

    return (left || right)
      ? (
        <div className={cls} ref={(wrap) => { this.wrap = wrap; }}>
          {this.renderButtons(left, 'left')}
          {this.renderButtons(right, 'right')}
          <Drag
            onDragStart={this.onDragStart}
            onDragMove={this.onDragMove}
            onDragEnd={this.onDragEnd}
          >
            <div className={`${prefixCls}-content`} style={style}>
              {children}
            </div>
          </Drag>
        </div>
      )
      : children;
  }
}
