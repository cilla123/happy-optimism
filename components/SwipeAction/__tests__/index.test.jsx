import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import SwipeAction from '../index';
import Button from '../../Button/index';

describe('SwipeAction', () => {
  it('renders correctly', () => {
    jest.useFakeTimers();
    const wrapper = mount(
      <SwipeAction
        left={[
          <Button theme="primary" onClick={jest.fn()}>左按钮1</Button>,
          <Button theme="warning" onClick={jest.fn()}>左按钮2</Button>,
        ]}
        right={[
          <Button theme="primary" onClick={jest.fn()}>右按钮1</Button>,
          <Button theme="warning" onClick={jest.fn()}>右按钮2</Button>,
        ]}>
        <div>左右都能滑动</div>
      </SwipeAction>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    jest.runAllTimers();
    wrapper.unmount();
  });

  it('left only', () => {
    const wrapper = render(
      <SwipeAction
        left={[
          <Button theme="primary" onClick={jest.fn()}>左按钮1</Button>,
          <Button theme="warning" onClick={jest.fn()}>左按钮2</Button>,
        ]}>
        <div>右滑</div>
      </SwipeAction>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('right only', () => {
    const wrapper = render(
      <SwipeAction
        right={[
          <Button theme="primary" onClick={jest.fn()}>右按钮1</Button>,
          <Button theme="warning" onClick={jest.fn()}>右按钮2</Button>,
        ]}>
        <div>左滑</div>
      </SwipeAction>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
