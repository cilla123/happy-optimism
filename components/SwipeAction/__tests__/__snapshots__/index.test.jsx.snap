// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeAction left only 1`] = `
<div
  class="za-swipeaction"
>
  <div
    class="za-swipeaction-actions-left"
  >
    <a
      aria-disabled="false"
      class="za-button theme-primary"
      role="button"
    >
      <span>
        左按钮1
      </span>
    </a>
    <a
      aria-disabled="false"
      class="za-button theme-warning"
      role="button"
    >
      <span>
        左按钮2
      </span>
    </a>
  </div>
  <div
    class="za-swipeaction-content"
    style="-webkit-transition-duration:undefinedms;transition-duration:undefinedms;-webkit-transform:translate3d(0px, 0, 0);transform:translate3d(0px, 0, 0)"
  >
    <div>
      右滑
    </div>
  </div>
</div>
`;

exports[`SwipeAction renders correctly 1`] = `
<SwipeAction
  animationDuration={300}
  left={
    Array [
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="primary"
      >
        左按钮1
      </Button>,
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="warning"
      >
        左按钮2
      </Button>,
    ]
  }
  moveDistanceRatio={0.5}
  moveTimeSpan={300}
  offset={10}
  prefixCls="za-swipeaction"
  right={
    Array [
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="primary"
      >
        右按钮1
      </Button>,
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="warning"
      >
        右按钮2
      </Button>,
    ]
  }
>
  <div
    className="za-swipeaction"
  >
    <div
      className="za-swipeaction-actions-left"
    >
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        key="0"
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="primary"
      >
        <a
          aria-disabled={false}
          className="za-button theme-primary"
          onClick={[Function]}
          onTouchStart={[Function]}
          role="button"
        >
          <span>
            左按钮1
          </span>
        </a>
      </Button>
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        key="1"
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="warning"
      >
        <a
          aria-disabled={false}
          className="za-button theme-warning"
          onClick={[Function]}
          onTouchStart={[Function]}
          role="button"
        >
          <span>
            左按钮2
          </span>
        </a>
      </Button>
    </div>
    <div
      className="za-swipeaction-actions-right"
    >
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        key="0"
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="primary"
      >
        <a
          aria-disabled={false}
          className="za-button theme-primary"
          onClick={[Function]}
          onTouchStart={[Function]}
          role="button"
        >
          <span>
            右按钮1
          </span>
        </a>
      </Button>
      <Button
        active={false}
        block={false}
        bordered={false}
        disabled={false}
        focus={false}
        key="1"
        loading={false}
        onClick={[Function]}
        prefixCls="za-button"
        theme="warning"
      >
        <a
          aria-disabled={false}
          className="za-button theme-warning"
          onClick={[Function]}
          onTouchStart={[Function]}
          role="button"
        >
          <span>
            右按钮2
          </span>
        </a>
      </Button>
    </div>
    <Drag
      onDragEnd={[Function]}
      onDragMove={[Function]}
      onDragStart={[Function]}
    >
      <div
        className="za-swipeaction-content"
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "WebkitTransform": "translate3d(0px, 0, 0)",
            "WebkitTransitionDuration": "undefinedms",
            "transform": "translate3d(0px, 0, 0)",
            "transitionDuration": "undefinedms",
          }
        }
      >
        <div>
          左右都能滑动
        </div>
      </div>
    </Drag>
  </div>
</SwipeAction>
`;

exports[`SwipeAction right only 1`] = `
<div
  class="za-swipeaction"
>
  <div
    class="za-swipeaction-actions-right"
  >
    <a
      aria-disabled="false"
      class="za-button theme-primary"
      role="button"
    >
      <span>
        右按钮1
      </span>
    </a>
    <a
      aria-disabled="false"
      class="za-button theme-warning"
      role="button"
    >
      <span>
        右按钮2
      </span>
    </a>
  </div>
  <div
    class="za-swipeaction-content"
    style="-webkit-transition-duration:undefinedms;transition-duration:undefinedms;-webkit-transform:translate3d(0px, 0, 0);transform:translate3d(0px, 0, 0)"
  >
    <div>
      左滑
    </div>
  </div>
</div>
`;
